File: product/conference/src/main/java/ee/ctob/api/error/ExceptionAdvice.java
package ee.ctob.api.error;


import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.ResponseStatus;

@Slf4j
@ControllerAdvice
public class ExceptionAdvice {

   @ExceptionHandler
   @ResponseStatus(HttpStatus.BAD_REQUEST)
   @ResponseBody
   protected ErrorResponse processException(BadRequestException ex) {
      log.warn(ex.getMessage(), ex);
      return ex.getError();
   }

   @ExceptionHandler
   @ResponseStatus(HttpStatus.FORBIDDEN)
   @ResponseBody
   protected ErrorResponse processException(ForbiddenException ex) {
      log.warn(ex.getMessage(), ex);
      return ex.getError();
   }

   @ExceptionHandler
   @ResponseStatus(HttpStatus.PRECONDITION_FAILED)
   @ResponseBody
   protected ErrorResponse processException(PreconditionsFailedException ex) {
      log.warn(ex.getMessage(), ex);
      return ex.getError();
   }

   @ExceptionHandler
   @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
   protected ErrorResponse processException(InternalServerErrorException ex) {
      log.error(ex.getMessage(), ex);
      return ex.getError();
   }
}

File: product/conference/src/main/kotlin/ee/ctob/api/error/ExceptionAdvice.kt
package ee.ctob.api.error

import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.ResponseStatus
import org.slf4j.LoggerFactory

@ControllerAdvice
class ExceptionAdvice {

    private val log = LoggerFactory.getLogger(ExceptionAdvice::class.java)

    @ExceptionHandler(BadRequestException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ResponseBody
    fun processException(ex: BadRequestException): ErrorResponse {
        log.warn(ex.message, ex)
        return ex.error
    }

    @ExceptionHandler(PreconditionsFailedException::class)
    @ResponseStatus(HttpStatus.PRECONDITION_FAILED)
    @ResponseBody
    fun processException(ex: PreconditionsFailedException): ErrorResponse {
        log.warn(ex.message, ex)
        return ex.error
    }
}